### PR TITLE
[SPARK-36455][SS] Provide an example of complex session window via flatMapGroupsWithState

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to add a new example of complex sessionization, which leverages flatMapGroupsWithState.

### Why are the changes needed?

We have replaced an example of sessionization from flatMapGroupsWithState to native support of session window. Given there are still use cases on sessionization which native support of session window cannot cover, it would be nice if we can demonstrate such case. It will also be used as an example of flatMapGroupsWithState.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually tested. Example data is given in class doc.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
I probably need to convert the Scala example to Java one as well. Marking this as draft for now.

**Comment by HeartSaVioR** [comment]
Addressed Java port. This is now ready to review.

**Comment by HeartSaVioR** [comment]
cc. @viirya @xuanyuanking

**Comment by HeartSaVioR** [comment]
retest this, please

File: examples/src/main/scala/org/apache/spark/examples/sql/streaming/StructuredComplexSessionization.scala [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  def endTime: Timestamp = sortedEvents.last.endTimestamp
+
+  override def toString: String = {
+    s"SessionAcc(events: $events / sorted: $sortedEvents / " +
+      s"start time: $startTime / endTime: $endTime)"
+  }
+}
+
+/**
+ * User-defined data type representing the session information returned by mapGroupsWithState.
```
Comment (viirya) [inline examples/src/main/scala/org/apache/spark/examples/sql/streaming/StructuredComplexSessionization.scala:290]: mapGroupsWithState -> flatMapGroupsWithState?

File: examples/src/main/java/org/apache/spark/examples/sql/streaming/JavaStructuredComplexSessionization.java [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+ * The supported types are following:
+ * - NEW_EVENT
+ * - CLOSE_SESSION
+ * <p>
+ * This example focuses to demonstrate the complex sessionization which uses two conditions
+ * on closing session; conditions are following:
+ * - No further event is provided for the user ID within 5 seconds
+ * - An event having CLOSE_SESSION as event_type is provided for the user ID
+ * <p>
+ * Usage: StructuredComplexSessionization <hostname> <port>
```
Comment (viirya) [inline examples/src/main/java/org/apache/spark/examples/sql/streaming/JavaStructuredComplexSessionization.java:66]: StructuredComplexSessionization -> JavaStructuredComplexSessionization?

**Review by viirya — APPROVED** [review]
found a few typos. otherwise looks good.

File: examples/src/main/java/org/apache/spark/examples/sql/streaming/JavaStructuredComplexSessionization.java [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+ * <hostname> and <port> describe the TCP server that Structured Streaming
+ * would connect to receive data.
+ * <p>
+ * To run this on your local machine, you need to first run a Netcat server
+ * `$ nc -lk 9999`
+ * and then run the example
+ * `$ bin/run-example sql.streaming.StructuredComplexSessionization
```
Comment (viirya) [inline examples/src/main/java/org/apache/spark/examples/sql/streaming/JavaStructuredComplexSessionization.java:73]: StructuredComplexSessionization -> JavaStructuredComplexSessionization?

**Comment by HeartSaVioR** [comment]
@viirya Thanks for the quick review! I addressed the comments. Please take a look. Thanks!

**Review by viirya — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master/3.2.

**Pull request closed without merging**